### DISCUSSION TITLE
Shopify discount function dynamic products

### DISCUSSION THREAD

**Discussion by phamquyetthang** [Shopify/function-examples#546] [Feedback: Discounts APIs]
I'm writing an app that suggests users can buy a few more products and get a discount (bundle).

For example: When a user visits product A, I will have algorithms to recommend the trio of products B - C - D with a 10% discount. These 3 products B - C - D are not fixed and will follow the recomendation logic depending on each person, product,...

How can I implement discount function for this case?

**Comment by jharlap** [comment]
Love the idea! I'm assuming you have a privacy-suitable way to collect buyer information and send it to your backend to generate the recommendations, and treating that as entirely out of scope for this discussion.

If I were building this, I'd use a theme extension and/or checkout UI extension to gather the buyer/product/whatever data, call my backend service with those data to get the recommendation, and store the recommended discount as cart attributes. I'd have a Product Discount Function that reads the cart attribute(s) and generates the corresponding discounts. In order to ensure that a malicious buyer can't abuse my Function, I'd have my server generate a signature (HMAC) that I can validate in the Function. Note that validating the signature in the Function code may consume a lot of the instruction count allowed by Shopify Functions, so I'd implement this in a language that is very energy efficient, like Rust, rather than javascript.

**Reply by kmcgowan2000** [reply]
What if you wanted to also make it so that, if the user selects product B (with 10% discount), products A and B are then bundled together (so that the user must keep the full price product A in their cart in order to receive the discount on product B)?

**Reply by kmcgowan2000** [reply]
Also great response @jharlap, the attention to detail is very nice to see.

**Reply by jharlap** [reply]
Thanks, @kmcgowan2000 😄 This is one of the gaps in the system that we're exploring solutions for, so today the answer is "you can't, really." The two ways for this to fail:
1. A buyer uses dev tools to mess with the client-side extension, such as creating a cart that would secure them a large discount and then disabling/breaking the extension before removing some stuff from the cart. The cart attributes wouldn't be updated to match the changed cart, and your Function would need to know that tampering has happened to avoid proposing the discount(s) you stored in the cart attributes. This can be solved by including the info that matters as part of the input to the signature, so I don't consider it a significant issue.
2. A merchant allows your personalized discount to combine with other discounts. As long as we don't provide a way for a Function's proposed discounts to express additional conditions (eg. "20% off product B if and only if product A has no discount"), unless the personalized discount cannot be combined with any other discount in the cart, then a merchant would have to be very careful to avoid overdiscounting and effectively selling products at a loss.

**Reply by kmcgowan2000** [reply]
Got it, this makes sense – for 1 of course the signature is a valid solution, and 2 was going to be my next question! I'll keep an eye out for when you land on a solution, but for now thanks for the clarification @jharlap 😄
